persist/fs/commitlog/reader.go

```diff
@@ -109,58 +165,171 @@ func (r *reader) Read() (
 	annotation ts.Annotation,
 	resultErr error,
 ) {
-	entry, err := r.readEntry()
-	if err != nil {
-		resultErr = err
-		return
+	if r.nextIndex == 0 {
+		go r.readLoop()
+		for i, decoderBuf := range r.decoderBufs {
+			go r.decoderLoop(decoderBuf, r.outBufs[i])
+		}
 	}
+	// Data is written into these channels in round-robin fashion, so reading them
+	// one at a time in the same order results in an ordered stream.
+	rr, ok := <-r.outBufs[r.nextIndex%r.numConc]
+	if !ok {
+		return Series{}, ts.Datapoint{}, xtime.Unit(0), ts.Annotation(nil), io.EOF
+	}
+	r.nextIndex++
+	return rr.series, rr.datapoint, rr.unit, rr.annotation, rr.resultErr
+}
 
-	metadata, ok := r.metadataLookup[entry.Index]
-	if !ok && len(entry.Metadata) != 0 {
-		r.metadataDecoder.Reset(entry.Metadata)
-		decoded, err := r.metadataDecoder.DecodeLogMetadata()
-		if err != nil {
-			resultErr = err
+func (r *reader) readLoop() {
+	index := 0
+	eofFound := false
+	for {
+		select {
+		case <-r.cancelCtx.Done():
+			for _, decoderBuf := range r.decoderBufs {
+				close(decoderBuf)
+			}
+			r.shutdownChan <- r.close()
 			return
+		default:
+			if eofFound {
+				continue
```

Comment:
Doesn't this cause an infinite loop?

Reply:
yeah technically, but it doesn't really matter because it ends when you call Close() and its in a background routine, but you're right its poor form. I found a way to make it not an infinite loop and I'll change it.

Reply:
Sounds good